### PR TITLE
feat(server): add support for multiple auth tokens via env vars

### LINKED ISSUE #338: Loading Auth Tokens from a file
I am using [agenix](https://github.com/ryantm/agenix) to encrypt my secrets. Agenix only exposes the path to the decrypted file, and not the content itself, necessitating to give an application the path to the file at runtime.

Since rustypaste either loads the auth token from the `AUTH_TOKEN` env var or the configuration file, I find myself unable to pass multiple auth tokens to the server.

I'd be interested in adding the parsing of an authentication file at server start, this authentication file could be provided via an `AUTH_FILE` env var, which holds the path, and the file itself could have a newline separated list of authentication tokens.

If there is a different or better method of implementing this feature I'd gladly hear about it.

### PR DESCRIPTION
**Pull request by Nydragon**
## Description

Add support for reading a list of newline separated tokens from a file
which path is being supplied through the following environment
variables:

- AUTH_FILE
- DELETE_FILE

Also add support for setting environment variables inside of fixtures through the optional `custom_env` function.

Also describe this new functionality in the readme.

## Motivation and Context

Need a way to supply multiple auth token akin to `AUTH_TOKEN` due to agenix behaviour.

closes #338

## How Has This Been Tested?

Added UTs for file content parsing. File reading has not received any UT, but has been manually tested.

### Added

- Multiple authentication and deletion token parsing via a file to which `AUTH_FILE` and `DELETION_FILE` point. This file needs to be a newline separated list of strings. Whitespace is discarded.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by tessus — APPROVED** [review]
Thanks for this PR. LGTM (unless I missed some Rust specific optimization I am not aware of, since I am not a Rust expert).

**Comment by Nydragon** [comment]
> Oh, could you add a test to the fixtures as well?

Sure! I didn't see an option to set env vars tho... Did i miss something?

**Comment by tessus** [comment]
> I was wondering if it would be better to `.trim()` the entire file or the individual lines after splitting (current behaviour).

One would have to run perf tests for the difference, but I doubt it is huge for a small to medium sized file with tokens.
Personally I like the current behavior better, since you took care of a few edge cases as well.

The only workflow optimization I can think of is that the file is not read and parsed every time a request is made, but only when the config is read or reloaded via `refresh_rate`.

If you want to try that, please go ahead.

**Comment by Nydragon** [comment]
> > I was wondering if it would be better to `.trim()` the entire file or the individual lines after splitting (current behaviour).
> 
> One would have to run perf tests for the difference, but I doubt it is huge for a small to medium sized file with tokens. Personally I like the current behavior better, since you took care of a few edge cases as well.
> 
> The only workflow optimization I can think of is that the file is not read and parsed every time a request is made, but only when the config is read or reloaded via `refresh_rate`.
> 
> If you want to try that, please go ahead.

That'd make definitely sense! I was a bit surprised to see that `get_tokens` is called on every request. Why is that? Wouldn't it be enough to just call the function periodically according to `refresh_rate`?

**Comment by tessus** [comment]
> That'd make definitely sense! I was a bit surprised to see that `get_tokens` is called on every request. Why is that? Wouldn't it be enough to just call the function periodically according to `refresh_rate`?

This is a good question. I think it developed this way. At the beginning there was only one token in the config file and one env var. I added the multiple tokens and delete tokens, and then it was refactored by somebody else to be used as a middleware with web grants.
In all these stages only the config was referenced (not the file read) and the ENV vars read, thus no real overhead.

File: src/config.rs [hunk 1]
```diff
@@ -170,6 +180,16 @@ impl Config {
                 if let Ok(env_token) = env::var(DELETE_TOKEN_ENV) {
                     tokens.insert(env_token);
                 }
+
+                if let Ok(env_path) = env::var(DELETE_FILE_ENV) {
+                    match read_to_string(&env_path) {
+                        Ok(s) => push_token_file(s, &mut tokens),
```
Comment (orhun) [inline src/config.rs:186]: No need for a function call:

```suggestion
                        Ok(s) => {
                            s.lines().filter(|l| !l.trim().is_empty()).for_each(|l| {
                                tokens.insert(l.to_string());
                            });
                        }
```

File: src/config.rs [hunk 2]
```diff
@@ -170,6 +180,16 @@ impl Config {
...
+                        Err(e) => {
+                            error!("Failed reading from deletion file {env_path}.\n{e}");
```
Comment (orhun) [inline src/config.rs:188]:
```suggestion
                            error!("failed to read deletion tokens from file ({env_path}) ({e})");
```

File: src/config.rs [hunk 3]
```diff
@@ -162,6 +163,15 @@ impl Config {
                 if let Ok(env_token) = env::var(AUTH_TOKEN_ENV) {
                     tokens.insert(env_token);
                 }
+                if let Ok(env_path) = env::var(AUTH_FILE_ENV) {
+                    match read_to_string(&env_path) {
+                        Ok(s) => push_token_file(s, &mut tokens),
```
Comment (orhun) [inline src/config.rs:168]: No need for a function call.

```suggestion
                        Ok(s) => {
                            s.lines().filter(|l| !l.trim().is_empty()).for_each(|l| {
                                tokens.insert(l.to_string());
                            });
                        }
```

File: src/config.rs [hunk 4]
```diff
@@ -203,6 +223,18 @@ impl Config {
...
+        .split('\n')
+        .filter_map(|s| match s.trim() {
+            "" => None,
+            s => Some(s),
+        })
+        .for_each(|t| {
+            tokens.insert(t.to_string());
+        });
+}
+
```
Comment (orhun) [inline src/config.rs:237]: Not needed.

**Review by orhun — COMMENTED** [review]
Looks good! We can simplify it a bit :)

**Comment by orhun** [comment]
> are you ok with adding another hook to add env vars to the fixtures?

Hmm, yeah, if it is not much of a hassle. Bash is annoying.

>  I was a bit surprised to see that get_tokens is called on every request. Why is that? 

We use a library called `actix_web_grants` which works by adding a middleware and then you can simply protect your endpoints with an annotation.

https://github.com/orhun/rustypaste/blob/1d5a9c60044ab0ece53868eccbe7cf011d0c5633/src/server.rs#L417

https://github.com/orhun/rustypaste/blob/1d5a9c60044ab0ece53868eccbe7cf011d0c5633/src/server.rs#L153

Implemented in #199 

That is why `extract_tokens` is called on every request since there is a guard. We can maybe cache some parts to make it more performant.

> Wouldn't it be enough to just call the function periodically according to refresh_rate?

That means we need to store the tokens somewhere (probably in memory) and that's not how the grants library works.

File: src/config.rs (hunk 1 above)
Comment (Nydragon) [reply src/config.rs:186]: I originally created a function to be able to write a UT for the parsing, is just a fixture fine then ?

**Comment by Nydragon** [comment]
Alright, thank you for the explanation, should I then leave out the improvements mentioned by tessus (https://github.com/orhun/rustypaste/pull/339#issuecomment-2304652210)?

I'll also add the env handling to the fixture script.

File: src/config.rs (hunk 1 above)
Comment (orhun) [reply src/config.rs:186]: Yup, it should be enough.

**Comment by Nydragon** [comment]
Should be all good now:

- added fixture for the file based authentication process
- added fixture for the file based deletion process
- added env var setting through the `rp_env` function, teardown is handled by putting `run_fixture` into a subshell
- applied code review requests

File: fixtures/README.md [hunk 5]
```diff
@@ -28,6 +28,11 @@ test-file-upload/
 ```sh
 #!/usr/bin/env bash
 
+# Optional
+rp_env() {
```
Comment (orhun) [inline fixtures/README.md:32]: I think it would be better if we rename this to `custom_env`.

File: fixtures/README.md [hunk 6]
```diff
@@ -28,6 +28,11 @@ test-file-upload/
...
+  # setting env variables such as AUTH_TOKEN or AUTH_FILE
```
Comment (orhun) [inline fixtures/README.md:33]: nit:
```suggestion
  # setting environment variables such as AUTH_TOKEN or AUTH_FILE
```

File: fixtures/test-fixtures.sh [hunk 7]
```diff
@@ -9,11 +9,14 @@ NC='\033[0m'
 run_fixture() {
   cd "$FIXTURE_DIR/$1" || exit 1
   source "test.sh"
+  # Check if rp_env exists
```
Comment (orhun) [inline fixtures/test-fixtures.sh:12]: it's clear, no need for a comment here :)

File: fixtures/test-fixtures.sh [hunk 8]
```diff
@@ -24,7 +27,9 @@ run_fixture() {
 
 main() {
   find * -maxdepth 0 -type d -print0 | while IFS= read -r -d '' fixture; do
-    run_fixture "$fixture"
+    # Since we are creating a subshell, all env vars created by rp_env will be lost
+    # Return code is preserved
+    (run_fixture "$fixture")
```
Comment (orhun) [inline fixtures/test-fixtures.sh:32]: Is there any side effects of this? Are we sure that the return code is preserved?

I'd test it manually with a failing fixture on purpose to see if the behavior is the same.

also nit:

```suggestion
    # Since we are creating a subshell, all environment variables created by rp_env will be lost
    # Return code is preserved
    (run_fixture "$fixture")
```

Comment (tessus) [reply fixtures/test-fixtures.sh:32]: Interesting. I would have used `custom_env_teardown()` and added it to `run_fixture()` before retuning.

But if the subshell works, it's certainly a better option.

Comment (Nydragon) [reply fixtures/test-fixtures.sh:32]: I echo'd the return value during my modifications and it did preserve the return code properly. As for any side effects I couldn't notice any.

<details><summary>Fixtures run normally:</summary>
<p>

image
</p>
</details> 

<details><summary>Modified `test` to fail</summary>
<p>

image

</p>
</details>

**Comment by Nydragon** [comment]
Applied all requests

**Review by tessus — COMMENTED** [review]
Great, thanks.

One last thing. Could you please update the README in the fixtures dir?

Can you please add the new function to the file format?

**Comment by Nydragon** [comment]
I think I already did that: https://github.com/orhun/rustypaste/pull/339/files#diff-3374186a79a0a6316047cb823494f238bdc644489b1b709bf65c192b5c05183a

**Comment by tessus** [comment]
Thanks, I missed it. All good then! I think it is ready to be merged.

File: src/lib.rs [hunk 9]
```diff
@@ -41,5 +41,11 @@ pub const CONFIG_ENV: &str = "CONFIG";
...
 pub const AUTH_TOKEN_ENV: &str = "AUTH_TOKEN";
 
+/// Environment variable for the path to a file containing multiple authentication token.
+pub const AUTH_FILE_ENV: &str = "AUTH_FILE";
+
 /// Environment variable for setting the deletion token.
 pub const DELETE_TOKEN_ENV: &str = "DELETE_TOKEN";
+
+/// Environment variable for the path to a file containing multiple deletion token.
+pub const DELETE_FILE_ENV: &str = "DELETE_FILE";
```
Comment (orhun) [inline src/lib.rs:51]: I was about to merge this but then had an epiphany, `DELETE_FILE` sounds more like "delete this file" rather than "a file that contains the deletion tokens". I think we should grammatically fix this while we have the chance :) So I propose:

- `AUTH_FILE` -> `AUTH_TOKENS_FILE`
- `DELETE_FILE` -> `DELETE_TOKENS_FILE`

Thoughts? @Nydragon @tessus

Comment (Nydragon) [reply src/lib.rs:51]: Completely agree, actually chose that name at the beginning, but replaced it with the current one because it was shorter.

Will update that as soon as I'm back home!

Comment (tessus) [reply src/lib.rs:51]: Shorter names only make sense, if it has to be used all the time....

Comment (Nydragon) [reply src/lib.rs]: Changed the names

**Review by orhun — APPROVED** [review]
Perfect, thanks!
